File: src/main/java/com/maximebertheau/emoji/EmojiLoader.java
package com.maximebertheau.emoji;

import org.json.JSONArray;
import org.json.JSONObject;

import java.io.*;
import java.util.ArrayList;
import java.util.Collections;
import java.util.LinkedList;
import java.util.List;

/**
 * Loads the emojis from a JSON database.
 *
 * @author Vincent DURMONT [vdurmont@gmail.com]
 */
public class EmojiLoader {
    /**
     * No need for a constructor, all the methods are static.
     */
    private EmojiLoader() {
    }

    /**
     * Loads a JSONArray of emojis from an InputStream, parses it and returns the
     * associated list of {@link Emoji}s
     *
     * @param stream the stream of the JSONArray
     * @return the list of {@link Emoji}s
     * @throws IOException if an error occurs while reading the stream or parsing
     *                     the JSONArray
     */
    public static List<Emoji> loadEmojis(InputStream stream) throws IOException {
        JSONArray emojisJSON = new JSONArray(inputStreamToString(stream));
        List<Emoji> emojis = new ArrayList<Emoji>(emojisJSON.length());
        for (int i = 0; i < emojisJSON.length(); i++) {

            JSONObject emojiObject = emojisJSON.getJSONObject(i);

            if (!emojiObject.has("unified")) continue;

            List<String> possibleUnicodes = new LinkedList<String>();
            possibleUnicodes.add(emojiObject.getString("unified"));

            if (emojiObject.has("variations")) {
                JSONArray variations = emojiObject.getJSONArray("variations");
                for (int j = 0; j < variations.length(); j++) {
                    possibleUnicodes.add(variations.getString(j));
                }
            }

            Collections.reverse(possibleUnicodes);

            for (String unicode : possibleUnicodes) {
                emojiObject.put("unified", unicode);
                Emoji emojiVariant = buildEmojiFromJSON(emojiObject);
                if (emojiVariant != null) {
                    emojis.add(emojiVariant);
                }
            }
        }
        return emojis;
    }

    private static String inputStreamToString(InputStream stream) throws IOException {
        StringBuilder sb = new StringBuilder();
        InputStreamReader isr = new InputStreamReader(stream, "UTF-8");
        BufferedReader br = new BufferedReader(isr);
        String read;
        while ((read = br.readLine()) != null) {
            sb.append(read);
        }
        br.close();
        return sb.toString();
    }

    protected static Emoji buildEmojiFromJSON(JSONObject json) throws UnsupportedEncodingException {
        if (!json.has("unified")) {
            return null;
        }

        byte[] bytes = extractEmojiFromUnified(json.getString("unified")).getBytes("UTF-8");

        boolean supportsFitzpatrick = json.has("skin_variations");
        List<String> aliases = jsonArrayToStringList(json.getJSONArray("short_names"));

        Category category = Category.parse(json.optString("category"));

        return new Emoji(json.optString("name"), supportsFitzpatrick, aliases, json.has("obsoleted_by"), category, json.optInt("sort_order"), bytes);
    }

    private static List<String> jsonArrayToStringList(JSONArray array) {
        List<String> strings = new ArrayList<String>(array.length());
        for (int i = 0; i < array.length(); i++) {
            strings.add(array.getString(i));
        }
        return strings;
    }

    private static String extractEmojiFromUnified(String unified) {
        String[] unifieds = unified.split("-");
        StringBuilder emoji = new StringBuilder();
        for (int i = 0; i < unifieds.length; i++) {
            int intValue = Integer.parseInt(unifieds[i], 16);
            char[] chars = Character.toChars(intValue);

            emoji.append(new String(chars));
        }

        return emoji.toString();
    }
}

File: src/main/java/com/maximebertheau/emoji/EmojiLoader.kt
package com.maximebertheau.emoji

import org.json.JSONArray
import org.json.JSONObject
import java.io.IOException
import java.io.UnsupportedEncodingException

internal object EmojiLoader {

    private const val PATH = "/emojis.json"

    /**
     * Loads a JSONArray of emojis from an InputStream, parses it and returns the
     * associated list of [Emoji]s
     *
     * @param stream the stream of the JSONArray
     * @return the list of [Emoji]s
     * @throws IOException if an error occurs while reading the stream or parsing
     * the JSONArray
     */
    @Throws(IOException::class)
    internal fun loadEmojis(): Sequence<Emoji> {
        val stream = EmojiLoader::class.java.getResourceAsStream(PATH)

        val json = try {
            stream.use { it.bufferedReader(Charsets.UTF_8).readText() }
        } catch (e: IOException) {
            throw RuntimeException(e)
        }

        return JSONArray(json).objects().asSequence()
                .filter { it.has("unified") }
                .flatMap { emojiObject ->
                    val variations = sequence<String> {
                        yield(emojiObject.getString("unified"))
                        yieldAll(emojiObject.optJSONArray("variations")?.strings().orEmpty())
                    }.reversed()

                    variations.map { variation ->
                        emojiObject to variation
                    }
                }
                .flatMap { (emojiObject, variation) ->
                    emojiObject.put("unified", variation)
                    emojiObject.toEmojis()
                }
    }

    @JvmStatic
    @Throws(UnsupportedEncodingException::class)
    internal fun JSONObject.toEmojis(): Sequence<Emoji> {
        val unified = optString("unified") ?: return emptySequence()
        val aliases = optJSONArray("short_names")?.strings() ?: return emptySequence()
        val category = optString("category")?.let(Category.Companion::parse) ?: return emptySequence()
        val name = optString("name")
        val isObsolete = has("obsoleted_by")
        val sortOrder = optInt("sort_order")

        val skinVariations = optJSONArray("skin_variations")?.arrays()
                ?.map {
                    val types = it.getString(0).split('-').map(SkinVariationType.Companion::fromUnified)
                    val variation = it.getString(1)

                    SkinVariation(types, variation)
                }
                .orEmpty()

        val emojiBase = Emoji(name, unified, aliases, isObsolete, category, sortOrder, skinVariations, isPristine = true)

        return sequence {
            yield(emojiBase)

            skinVariations.forEach { variation ->
                val newAliases = emojiBase.aliases.map { alias ->
                    if (variation.types.isEmpty()) {
                        alias
                    } else {
                        alias + ":" + variation.types.joinToString(separator = "") { ":${it.alias}:" }.trimEnd(':')
                    }
                }
                yield(emojiBase.copy(aliases = newAliases, unified = variation.unified, isPristine = false, skinVariations = emptyList<SkinVariation>()))
            }
        }
    }

    private fun JSONArray.objects() = (0 until length()).map(::getJSONObject)
    private fun JSONArray.arrays() = (0 until length()).map(::getJSONArray)
    private fun JSONArray.strings() = (0 until length()).map(::getString)
    private fun <T> Sequence<T>.reversed() = toList().asReversed().asSequence()
}